plugins/node/opentelemetry-instrumentation-dns/src/dns.ts

```diff
@@ -0,0 +1,210 @@
+    return function patchedLookup(
+      this: {},
+      hostname: string,
+      ...args: unknown[]
+    ) {
+      if (
+        utils.isIgnored(hostname, plugin._config.ignoreHostnames, (e: Error) =>
+          plugin._logger.error('caught ignoreHostname error: ', e)
+        )
+      ) {
+        return original.apply(this, [hostname, ...args]);
+      }
+
+      const argsCount = args.length;
+      plugin._logger.debug('wrap lookup callback function and starts span');
+      const name = utils.getOperationName('lookup');
+      const span = plugin._startDnsSpan(name, {
+        attributes: {
+          [AttributeNames.PEER_HOSTNAME]: hostname,
+        },
+      });
+
+      const originalCallback = args[argsCount - 1];
+      if (typeof originalCallback === 'function') {
+        args[argsCount - 1] = plugin._wrapLookupCallback(
+          originalCallback,
+          args[argsCount - 2],
+          span
+        );
+        return plugin._safeExecute(span, () =>
+          (original as LookupFunctionSignature).apply(this, [
+            hostname,
+            ...args,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          ] as any)
+        );
+      } else {
+        const promise = plugin._safeExecute(span, () =>
+          (original as LookupPromiseSignature).apply(this, [hostname, ...args])
+        );
+        promise.then(
+          result => {
+            utils.setLookupAttributes(span, result as LookupAddress);
+            span.end();
+          },
+          (e: NodeJS.ErrnoException) => {
+            utils.setError(e, span, process.version);
+            span.end();
+          }
+        );
+
+        return promise;
+      }
+    };
+  }
+
+  /**
+   * Start a new span with default attributes and kind
+   */
+  private _startDnsSpan(name: string, options: Omit<SpanOptions, 'kind'>) {
+    return this.tracer
+      .startSpan(name, { ...options, kind: SpanKind.CLIENT })
+      .setAttribute(
+        AttributeNames.COMPONENT,
+        '@opentelemetry/instrumentation-dns'
+      );
+  }
+
+  /**
+   * Wrap lookup callback function
+   */
+  private _wrapLookupCallback(
+    original: Function,
+    options: unknown,
+    span: Span
+  ): LookupCallbackSignature {
+    const plugin = this;
+    return function wrappedLookupCallback(
+      this: {},
+      err: NodeJS.ErrnoException | null,
+      address: string | LookupAddress[],
+      family?: AddressFamily
+    ): void {
+      plugin._logger.debug('executing wrapped lookup callback function');
+
+      if (err !== null) {
+        utils.setError(err, span, process.version);
+      } else {
+        utils.setLookupAttributes(span, address, family);
+      }
+
+      span.end();
+      plugin._logger.debug('executing original lookup callback function');
+      return original.apply(this, arguments);
+    };
+  }
+
+  /**
+   * Safely handle "execute" callback
+   */
+  private _safeExecute<T extends (...args: unknown[]) => ReturnType<T>>(
```

Comment:
Could you use https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-instrumentation/src/utils.ts#L24 instead ?